pkg/cloudprovider/aws/amifamily/bottlerocket.go

```diff
@@ -0,0 +1,100 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package amifamily
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/cloudprovider"
+	"github.com/aws/karpenter/pkg/cloudprovider/aws/apis/v1alpha1"
+	"github.com/aws/karpenter/pkg/utils/functional"
+	"github.com/aws/karpenter/pkg/utils/injection"
+)
+
+func init() {
+	bottlerocket := Bottlrocket{}
+	amiFamilies[bottlerocket.Name()] = bottlerocket
+}
+
+type Bottlrocket struct{}
+
+// Name is the name of the AMI Family
+func (b Bottlrocket) Name() string {
+	return v1alpha1.AMIFamilyBottlerocket
+}
+
+// SSMAlias returns the AMI Alias to query SSM
+func (b Bottlrocket) SSMAlias(version string, instanceType cloudprovider.InstanceType) string {
+	arch := "x86_64"
+	amiSuffix := ""
+	if !instanceType.NvidiaGPUs().IsZero() {
+		amiSuffix = "-nvidia"
+	}
+	if instanceType.Architecture() == v1alpha5.ArchitectureArm64 {
+		arch = instanceType.Architecture()
+	}
+	return fmt.Sprintf("/aws/service/bottlerocket/aws-k8s-%s%s/%s/latest/image_id", version, amiSuffix, arch)
+}
+
+// UserData returns the default userdata script for the AMI Family
+func (b Bottlrocket) UserData(ctx context.Context, constraints *v1alpha1.Constraints, instanceTypes []cloudprovider.InstanceType, additionalLabels map[string]string, caBundle *string) string {
+	userData := fmt.Sprintf("[settings.kubernetes]\n\"cluster-name\" = \"%s\"\n\"api-server\" = \"%s\"\n", injection.GetOptions(ctx).ClusterName, injection.GetOptions(ctx).ClusterEndpoint)
+	if len(constraints.KubeletConfiguration.ClusterDNS) > 0 {
+		userData += fmt.Sprintf("\"cluster-dns-ip\" = \"%s\"\n", constraints.KubeletConfiguration.ClusterDNS[0])
+	}
+	if caBundle != nil {
+		userData += fmt.Sprintf("\"cluster-certificate\" = \"%s\"\n", *caBundle)
+	}
+	nodeLabelArgs := functional.UnionStringMaps(additionalLabels, constraints.Labels)
+	if len(nodeLabelArgs) > 0 {
+		userData += "[settings.kubernetes.node-labels]\n"
+		for key, val := range nodeLabelArgs {
+			userData += fmt.Sprintf("\"%s\" = \"%s\"\n", key, val)
+		}
+	}
+	if len(constraints.Taints) > 0 {
+		userData += "[settings.kubernetes.node-taints]\n"
+		sorted := sortedTaints(constraints.Taints)
+		for _, taint := range sorted {
+			userData += fmt.Sprintf("\"%s\"=\"%s:%s\"\n", taint.Key, taint.Value, taint.Effect)
+		}
+	}
+	return base64.StdEncoding.EncodeToString([]byte(userData))
+}
+
+// BlockDeviceMappings returns the default block device mappings for the AMI Family
+func (b Bottlrocket) BlockDeviceMappings(constraints *v1alpha1.Constraints) []*v1alpha1.BlockDeviceMapping {
```

Comment:
This makes me think, do we want to expose a full launch template abstraction to handle some of this?

Will AMIFamilies have different defaults across the launch template? Seems like they do so far?

Reply:
I've looked through the other Launch Template Parameters and I believe the AMI Alias, UserData, and BlockDeviceMappings are the only ones that are dependent upon which AMI Family is used.